core/capabilities/remote/target/request/receiver_request.go

```diff
@@ -0,0 +1,222 @@
+	capabilityID     string
+	capabilityDonID  string
+
+	dispatcher types.Dispatcher
+
+	requesters              map[p2ptypes.PeerID]bool
+	responseSentToRequester map[p2ptypes.PeerID]bool
+
+	createdTime time.Time
+
+	response *response
+
+	callingDon commoncap.DON
+
+	requestMessageID string
+	requestTimeout   time.Duration
+
+	mux sync.Mutex
+}
+
+func NewReceiverRequest(capability capabilities.TargetCapability, capabilityID string, capabilityDonID string, capabilityPeerId p2ptypes.PeerID,
+	callingDon commoncap.DON, requestMessageID string,
+	dispatcher types.Dispatcher, requestTimeout time.Duration) *ReceiverRequest {
+	return &ReceiverRequest{
+		capability:              capability,
+		createdTime:             time.Now(),
+		capabilityID:            capabilityID,
+		capabilityDonID:         capabilityDonID,
+		capabilityPeerId:        capabilityPeerId,
+		dispatcher:              dispatcher,
+		requesters:              map[p2ptypes.PeerID]bool{},
+		responseSentToRequester: map[p2ptypes.PeerID]bool{},
+		callingDon:              callingDon,
+		requestMessageID:        requestMessageID,
+		requestTimeout:          requestTimeout,
+	}
+}
+
+func (e *ReceiverRequest) OnMessage(ctx context.Context, msg *types.MessageBody) error {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+
+	if msg.Sender == nil {
+		return fmt.Errorf("sender missing from message")
+	}
+
+	requester := remote.ToPeerID(msg.Sender)
+	if err := e.addRequester(requester); err != nil {
+		return fmt.Errorf("failed to add requester to request: %w", err)
+	}
+
+	if e.minimumRequiredRequestsReceived() && !e.hasResponse() {
+		if err := e.executeRequest(ctx, msg.Payload); err != nil {
+			e.setError(types.Error_INTERNAL_ERROR, err.Error())
+		}
+	}
+
+	if err := e.sendResponses(); err != nil {
+		return fmt.Errorf("failed to send responses: %w", err)
+	}
+
+	return nil
+}
+
+func (e *ReceiverRequest) Expired() bool {
+	return time.Since(e.createdTime) > e.requestTimeout
+}
+
+func (e *ReceiverRequest) Cancel(err types.Error, msg string) error {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+
+	if e.hasResponse() {
+		return fmt.Errorf("request already has response")
+	}
+
+	e.setError(err, msg)
+	if err := e.sendResponses(); err != nil {
+		return fmt.Errorf("failed to send responses: %w", err)
+	}
+
+	return nil
+}
+
+func (e *ReceiverRequest) executeRequest(ctx context.Context, payload []byte) error {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, e.requestTimeout)
+	defer cancel()
+
+	capabilityRequest, err := pb.UnmarshalCapabilityRequest(payload)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal capability request: %w", err)
+	}
+
+	capResponseCh, err := e.capability.Execute(ctxWithTimeout, capabilityRequest)
+
+	if err != nil {
+		return fmt.Errorf("failed to execute capability: %w", err)
+	}
+
+	// TODO working on the assumption that the capability will only ever return one response from its channel (for now at least)
+	capResponse := <-capResponseCh
```

Comment:
Could you maybe add a comment in OnMessage() explaining that once an execution starts, all subsequent incoming messages will be blocked until it finishes? It wasn't obvious to me at first and I thought a slow underlying execution can cause problems, but it seems to be fine.